src/xdp/program.c

```diff
@@ -11,10 +11,183 @@
 #include "programinspect.h"
 #include "program.tmh"
 
+typedef struct _EBPF_PROG_TEST_RUN_CONTEXT {
+    char* Data;
+    SIZE_T DataSize;
+} EBPF_PROG_TEST_RUN_CONTEXT;
+
 typedef struct _EBPF_XDP_MD {
     xdp_md_t Base;
+    EBPF_PROG_TEST_RUN_CONTEXT* ProgTestRunContext;
 } EBPF_XDP_MD;
 
+static __forceinline NTSTATUS EbpfResultToNtStatus(ebpf_result_t Result)
+{
+    switch (Result) {
+    case EBPF_SUCCESS:
+        return STATUS_SUCCESS;
+    case EBPF_INVALID_ARGUMENT:
+        return STATUS_INVALID_PARAMETER;
+    case EBPF_NO_MEMORY:
+        return STATUS_NO_MEMORY;
+    default:
+        return STATUS_UNSUCCESSFUL;
+    }
+}
+
+//
+// Routines for BPF_PROG_TEST_RUN.
+//
+
+static void EbpfProgramTestRunContextFree(_In_opt_ _Post_invalid_ EBPF_PROG_TEST_RUN_CONTEXT* Context)
+{
+    if (Context == NULL) {
+        return;
+    }
+
+    if (Context->Data != NULL) {
+        ExFreePool(Context->Data);
+    }
+    ExFreePool(Context);
+}
+
+/**
+ * @brief Build a EBPF_XDP_MD Context for the eBPF program. This includes copying the packet data and
+ * metadata into a contiguous buffer and building an MDL chain for the same.
+ *
+ * @param[in] DataIn The packet data.
+ * @param[in] DataSizeIn The size of the packet data.
+ * @param[in] context_in The Context.
+ * @param[in] ContextSizeIn The size of the Context.
+ * @param[out] Context The Context to be passed to the eBPF program.
+ * @retval STATUS_SUCCESS The operation was successful.
+ * @retval STATUS_INVALID_PARAMETER One or more parameters are incorrect.
+ * @retval STATUS_NO_MEMORY Failed to allocate resources for this operation.
+ */
+static ebpf_result_t
+XdpCreateContext(
```

Comment:
These routines need spin/fuzz testing via bpf_prog_test_run - please add some randomized input cases to spinxsk.c.

Reply:
Fixed.